### PR TITLE
Fix failing component auto-updates

### PR DESCRIPTION
**Pull request by mherrmann**
Resolves https://github.com/brave/brave-browser/issues/10464
Also resolves https://github.com/brave/brave-browser/issues/10280

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed).
- [ ] Requested a security/privacy review as needed.

## Reviewer Checklist:

- [x] New files have MPL-2.0 license header.
- [ ] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by mherrmann** [comment]
@jumde the implementation is complete from my side. Can you review, or can you suggest someone else to review? (It's only my second PR to brave-core.)

File: patches/components-component_updater-component_updater_service.cc.patch
Comment (jumde) [inline patches/components-component_updater-component_updater_service.cc.patch:9]: You should be able to do this without patches, see code in: `chrome/browser/extensions/extension_install_prompt.cc`

Comment (mherrmann) [reply patches/components-component_updater-component_updater_service.cc.patch]: Thank you very much for the suggestion. This technique was new to me, and it seems very elegant. Unfortunately, after trying to use it here (see my commits, esp. `Use #define to avoid having to create patches`), I'm not sure it will be feasible.

In `extension_install_prompt`, which you mentioned, the overwritten method `GetDialogTitle` is public:

    class Prompt {
     public:
      ...
      base::string16 GetDialogTitle() const;

Furthermore, and importantly, `GetDialogTitle` is only called from _outside_ `extension_install_prompt.cc`. This means that the macro that renames `GetDialogTitle` to `GetDialogTitle_ChromiumImpl` has no effect on other files. That is, other source files will call the _overwritten_ `GetDialogTitle` and not the original `GetDialogTitle_ChromiumImpl`.

In the present case of `component_updater_service.cc`, the callers of the overwritten method `CheckForUpdates` are unfortunately in the same file. That is, when I define a macro that renames it to `CheckForUpdates_ChromiumImpl`, then the call sites in that file are _also_ renamed. So even if I then define a new implementation of `CheckForUpdates`, that implementation won't be called because `CheckForUpdates_ChromiumImpl` is called instead.

I could fix this by copying method definitions until I get to the public entry point in this situation, which is the method `RegisterComponent`. Precisely: The public method `RegisterComponent` calls the private method `Start`, which in turn calls the private method `CheckForUpdates`:

    CrxUpdateService::RegisterComponent(...) {
        ...
        Start();
        ...
    }

    CrxUpdateService::Start(...) {
        ...
        CheckForUpdates();
        ...
    }

These are all in `component_updater_service.cc`, so the `CheckForUpdates()` call gets replaced by `CheckForUpdates_ChromiumImpl()`. What I would have to do is copy their implementations into `brave-core`. But that's quite a lot of code and very likely to break or become stale as Chromium is updated.

Can you advise what you think I should do @jumde?

Comment (jumde) [reply patches/components-component_updater-component_updater_service.cc.patch]: @mherrmann - What do you think about creating a chromium_src override for `UpdateClientImpl::Update` instead?

Comment (mherrmann) [reply patches/components-component_updater-component_updater_service.cc.patch]: @jumde I don't think it will work in `UpdateClientImpl::Update` because `CrxDataCallback` is a `OnceCallback`. (The obvious way to implement your suggestion would be to have a `for` loop in `UpdateClientImpl::Update` that submits multiple update tasks instead of a single one. But because of the `OnceCallback`, only the first of these tasks would be able to query the current components.)

After some more analysis, I think the best place is in `update_checker`, injected via `UpdateClientFactory`. I've committed an implementation of this. If you could be so kind to review this, that would be great. From my side, I would deem it merge-able. Before doing so, I would squash my commits and (if you think I should) remove the logging calls. Please let me know :-)

**Comment by mherrmann** [comment]
@jumde just so it's visible here too, instead of hidden in the [review conversation](https://github.com/brave/brave-core/pull/7164#discussion_r529280252): the PR is ready to merge from my side (excl. squash, rebase and potentially removing log statements). I'd kindly ask that you review please and let me know what you think.

**Comment by mherrmann** [comment]
@mkarolin @pes10k can I also request your review on this? (@jumde suggested I ask you.) The short summary is: When you open `brave://components` and wait for 1 minute, then you see a bunch of update errors on the page. The reason for this is explained in my committed file `update_checker.h`. This is only my second PR to brave-core, so I would love to hear your thoughts.

**Comment by jumde** [comment]
@mherrmann - This fix also installs components (during auto-update) that shouldn't be shipped with Brave like `Federated Learning Of Cohorts`, `Legacy TLS Deprecation Configuration`, `Subresource FIlter Rules`. Only the components that are installed at the first run, should be updated with auto-update.

**Comment by mherrmann** [comment]
@jumde do you have an idea why those components are being installed?

**Comment by jumde** [comment]
> @jumde do you have an idea why those components are being installed?

@mherrmann - I need to dig into the chromium code to see why this might be happening.

**Comment by mherrmann** [comment]
Okay, thank you. As a workaround, could we make go-update return "no update" responses for those components we do not wish to be installed?

**Comment by jumde** [comment]
> Okay, thank you. As a workaround, could we make go-update return "no update" responses for those components we do not wish to be installed?

I think it would be easy to fix this on brave-core. If we do it on go-update, this will require updating go-update every time a component is added to the chromium code and it adds a testing burden as well.

**Comment by mherrmann** [comment]
> I think it would be easy to fix this on brave-core. If we do it on go-update, this will require updating go-update every time a component is added to the chromium code and it adds a testing burden as well.

Ah, I see. That makes sense. Thank you. I'm very curious what you'll find about those additionally installed components.

**Comment by mherrmann** [comment]
@jumde I develop on Linux. I just started Brave Nightly v1.19.36 on a Windows VM I haven't touched in a month. I have the components you mentioned also on this Windows VM:

image

You wrote that my PR here installs these unwanted components. Do I presume correctly that you meant that it _updates_ these components, possibly from a non-existent skeleton implementation?

I'll dig a little further how these components get there in the first place.

**Comment by mherrmann** [comment]
After some further digging, I have a better understanding how components are installed. Chromium installs its components in `chrome/browser/component_updater/registration.cc`. This is called from `chrome_browser_main.cc` in `ChromeBrowserMainParts::PreMainMessageLoopRunImpl`. Brave has a subclass `BraveBrowserMainParts` of `ChromeBrowserMainParts` that however does not override `PreMainMessageLoopRunImpl`.

What I don't understand is that apparently there is a wish to exclude certain Google components from being shipped with Brave. But besides some extra Brave components, my components in Chrome and Brave are basically the same.

The closest thing I could find to an implementation is `IsBlacklisted` in brave-core, which gets invoked from `UserMayLoad`. However, the associated classes refer to extensions, not components. I'm not sure yet whether that logic is then also used for components.

**Comment by mherrmann** [comment]
@jumde kindly pointed me to https://github.com/brave/brave-browser/wiki/Brave-Components, which lists the components officially supported by Brave, and those officially unsupported.

What surprises me is that my local Brave installation has many of the components that are supposed to be unsupported. For example, "Certificate Error Assistant" and "Crowd Deny".

@jumde also pointed out that components can be excluded from Brave by overriding their `Register...` methods. For example, `RegisterSafetyTipsComponent`. Brave does override some such registration functions in `brave-core/chromium_src/chrome/browser/component_updater`. However, as far as I can see, it does not actually stop any components from being installed.

Coming back to this PR: It doesn't seem to me like it installs any components that shouldn't be there. But yes, it does update all components, including the unwanted ones.

If updating unwanted components is a problem, then this PR shouldn't be merged yet and we have to prevent the components from being there in the first place. Otherwise, I would tackle removing these unwanted components in the separate existing issue https://github.com/brave/brave-browser/issues/8709.

@jumde do you see an issue with merging this PR and tackling https://github.com/brave/brave-browser/issues/8709 separately?

**Comment by mherrmann** [comment]
After reading https://github.com/brave/brave-browser/issues/10607#issuecomment-685186576, I think I can answer my question myself. It probably would be a problem to merge this PR. The reason is that while Brave does not prevent Chromium from _registering_ unwanted components, it does not install them. My suspicion is that Chromium installs components by registering them with version `0.0.0.0`, then performing an automatic update check. In Brave, this automatic update check was previously broken (hence the present PR). Hence the components were not installed.

So yes, I do think that we should tackle https://github.com/brave/brave-browser/issues/8709 before merging this PR.

File: chromium_src/components/update_client/update_checker.cc
Comment (bridiver) [inline chromium_src/components/update_client/update_checker.cc:15]: please run clang format

File: chromium_src/components/update_client/update_checker.h
Comment (bridiver) [inline chromium_src/components/update_client/update_checker.h:11]: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: chromium_src/components/update_client/update_checker.cc
Comment (mkarolin) [inline chromium_src/components/update_client/update_checker.cc:40]: Haven't looked through the code much, but is this safe for multiple invocations? For example if an automatic update check is in progress and I got to brave://components page and click a button to check for updates?

File: chromium_src/components/update_client/update_checker.cc
Comment (mkarolin) [inline chromium_src/components/update_client/update_checker.cc:59]: Looks to me like this will result in a new thread (in `UpdateCheckerImpl::CheckForUpdates`) for each component in the ids list, is that a concern?

Comment (mkarolin) [reply chromium_src/components/update_client/update_checker.cc:59]: Or maybe I am wrong here since we only check for the next id when the result of the previous one becomes available.

File: chromium_src/components/update_client/update_checker.cc
Comment (mherrmann) [reply chromium_src/components/update_client/update_checker.cc]: Done in 65f682e.

File: chromium_src/components/update_client/update_checker.h
Comment (mherrmann) [reply chromium_src/components/update_client/update_checker.h]: Implemented in d43fe7e.

File: chromium_src/components/update_client/update_checker.h
Comment (mherrmann) [reply chromium_src/components/update_client/update_checker.h]: Made private in e74d017.

File: chromium_src/components/update_client/update_checker.cc
Comment (mherrmann) [reply chromium_src/components/update_client/update_checker.cc:40]: I think it's safe now, and believe it is also reasonably safe for the future.

Chromium [creates a new `UpdateChecker` for every update request](https://source.chromium.org/chromium/chromium/src/+/master:components/update_client/update_engine.cc;l=157;drc=52c253e300831067aa75506906a4af2064edadb2). So at the moment, when you perform an on-demand update check while an automatic one is in progress, then a new instance is created that doesn't touch the one that is already running. This applies both to Chromium's base implementation and the one in this PR.

From the way `UpdateChecker` is designed and currently used, it seems to me that the intention to have it use throwaway instances is pretty deeply ingrained. Of course, there are no guarantees. But I don't think it is likely to change.

The present implementation tries to introduce as little new code as possible by delegating to `UpdateChecker`. A result of this is that it does spawn a new thread for every request. But these threads are only spawned sequentially, and `SequentialUpdateChecker` hides this fact from its callers. It gathers results internally and fails or succeeds as a whole before returning to the caller. Because it uses several threads and multiple requests, it is more brittle. (If a single request fails, then the entire update check fails.) But I'm not sure there is a better way to work around this. We could perform all requests in a single thread. But then we'd have to duplicate quite a bit more code from `UpdateChecker`.

Comment (mkarolin) [reply chromium_src/components/update_client/update_checker.cc:40]: Thanks for the detailed info, @mherrmann! Threads are definitely not an issue. I initially didn't look in the right place and thought the code was creating a new thread for each request concurrently, not sequentially. My only concern was the possibility of reentrancy, but if that can't happen, then it's all good.

File: chromium_src/components/update_client/update_checker.cc
Comment (jumde) [inline chromium_src/components/update_client/update_checker.cc:45]: @mherrmann - To avoid the privacy risk of leaking components that the user has installed, it would help if the request for the next component update is initiated after a random delay. Check this: https://github.com/brave/brave-core/blob/612896e4b30cca7352dfe2aa1247e2995f97b580/vendor/bat-native-ledger/src/bat/ledger/internal/common/time_util.cc#L40

Comment (mherrmann) [reply chromium_src/components/update_client/update_checker.cc:45]: @jumde I'm sorry, I don't quite understand your comment. As far as I'm aware, the implementation that is currently in Brave already leaks the components that the user has installed. Also, I don't understand how a random delay could mitigate the issue (the update request would still say which components the user has). If I'm not wrong in saying that Brave already leaks the installed components (out of technical necessity), then this PR doesn't make matters worse and my personal preference would be to merge this PR and address your concerns in a follow-up issue. But of course, I'll be happy to be overruled by you or other people more senior at Brave than me.

**Comment by mherrmann** [comment]
Tested the latest CI build for this PR on Linux (brave-browser-nightly_1.20.47_amd64.deb). Works as expected for me. After 60 seconds on `brave://components`, I see:

image

- no more update failures.

The only question mark is the new "Hyphenation" component. It seems this was introduced in a recent Chromium version. If Brave does not want to include it, then we should block it as in https://github.com/brave/brave-core/pull/7331.

**Review by mihaiplesa — APPROVED** [review]
Tested on macOS and works as expected. CI passed on all.

**Comment by simonhong** [comment]
Kindly ping to @brave/patch-reviewers :)

**Comment by mihaiplesa** [comment]
@bridiver said this looks fine, just waiting for @jumde's last thoughts.

**Review by simonhong — APPROVED** [review]

File: chromium_src/components/update_client/update_checker.cc
Comment (jumde) [reply chromium_src/components/update_client/update_checker.cc:45]: @mherrmann - Randomized delay reduces linkability of that sequential requests for components for a given user.  If you think adding a random delay will be a bigger change, feel free to file a follow-up issue for this.

**Review by jumde — APPROVED** [review]
lgtm

**Comment by mherrmann** [comment]
I rebased onto `master` and force-pushed to get a CI build one last time before merging.

**Comment by mherrmann** [comment]
The post-init check fails because of what I believe is a bug in `lint.py`

```/home/ubuntu/workspace/pr-brave-browser-mherrmann-fix-google-components-autoupdate-post-init/src
10:07:48  > ninja -C /home/ubuntu/workspace/pr-brave-browser-mherrmann-fix-google-components-autoupdate-post-init/src/out/Component_audit brave:audit_deps
10:07:48  ninja: Entering directory `/home/ubuntu/workspace/pr-brave-browser-mherrmann-fix-google-components-autoupdate-post-init/src/out/Component_audit'
10:07:50  Error running format check:
10:07:50  Traceback (most recent call last):
10:07:50    File "/home/ubuntu/workspace/pr-brave-browser-mherrmann-fix-google-components-autoupdate-post-init/src/brave/build/commands/scripts/lint.py", line 35, in RunFormatCheck
10:07:50      if HasFormatErrors():
10:07:50    File "/home/ubuntu/workspace/pr-brave-browser-mherrmann-fix-google-components-autoupdate-post-init/src/brave/build/commands/scripts/lint.py", line 26, in HasFormatErrors
10:07:50      print(diff)
10:07:50  UnicodeEncodeError: 'ascii' codec can't encode characters in position 265353-265360: ordinal not in range(128)```

The person who recently committed the `print(diff)` line has been alerted. @mihaiplesa is also looking into this.

**Review by mihaiplesa — APPROVED** [review]

**Comment by mherrmann** [comment]
The post-init check succeeded now. I also tested the CI build on Linux and am getting the expected result. On `brave://components`, after 60 seconds, an automatic update check is performed and no more errors appear:

image

Merging this.

**Comment by mherrmann** [comment]
@bridiver or @mkarolin can I get your approval on this PR please? I can't merge without another (4th) approving review.

image

image

**Comment by mihaiplesa** [comment]
Merging as the last change was minor.

**Pull request merged**
